crates/biome_css_analyze/src/lint/nursery/use_sorted_properties.rs

```diff
@@ -0,0 +1,387 @@
+use std::{
+    borrow::Cow,
+    collections::{BTreeSet, HashSet},
+};
+
+use biome_analyze::{
+    context::RuleContext, declare_lint_rule, ActionCategory, Ast, FixKind, Rule, RuleAction,
+    RuleDiagnostic,
+};
+use biome_console::markup;
+use biome_css_syntax::{
+    AnyCssDeclarationName, AnyCssDeclarationOrRule, AnyCssProperty, AnyCssRule,
+    CssDeclarationOrRuleBlock, CssDeclarationOrRuleList, CssDeclarationWithSemicolon,
+    CssIdentifier, CssLanguage, CssSyntaxKind,
+};
+use biome_rowan::{AstNode, BatchMutationExt, NodeOrToken, SyntaxNode, TokenText};
+use biome_string_case::StrOnlyExtension;
+
+use crate::{
+    keywords::VENDOR_PREFIXES,
+    order::PROPERTY_ORDER_MAP,
+    utils::{get_longhand_sub_properties, get_reset_to_initial_properties, vender_prefix},
+    CssRuleAction,
+};
+
+declare_lint_rule! {
+    /// Enforce ordering of CSS properties.
+    ///
+    /// This rule checks if the properties and nested rules are in a consistent order.
+    ///
+    /// The expected ordering is roughly:
+    /// 1. Custom properties
+    /// 1. Layout properties (display, flex, grid)
+    /// 1. Margin & padding properties
+    /// 1. Typography properties (font, color)
+    /// 1. Interaction properties (pointer-events, visibility)
+    /// 1. Background & border properties
+    /// 1. Transition & animation properties
+    /// 1. Nested rules, media queries & other at-rules
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```css,expect_diagnostic
+    /// p { color: black; display: block; }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```css
+    /// p { display: block; color: black; }
+    /// ```
+    ///
+    pub UseSortedProperties {
+        version: "next",
+        name: "useSortedProperties",
+        language: "css",
+        recommended: true,
+        fix_kind: FixKind::Safe,
+    }
+}
+
+pub struct UseSortedPropertiesState {
+    block: CssDeclarationOrRuleBlock,
+    original_properties: Vec<AnyCssDeclarationOrRule>,
+    sorted_properties: RecessOrderProperties,
+    is_unsafe_to_sort: bool,
+}
+
+impl Rule for UseSortedProperties {
+    type Query = Ast<CssDeclarationOrRuleBlock>;
+    type State = UseSortedPropertiesState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Option<Self::State> {
+        let query_result = ctx.query();
+
+        let original_properties = query_result
+            .items()
+            .into_iter()
+            .collect::<Vec<AnyCssDeclarationOrRule>>();
+        let is_unsafe_to_sort = contains_shorthand_after_longhand(&original_properties)
+            || contains_unknown_property(&original_properties);
```

Comment:
If `is_unsafe_to_sort` is `true` we can bail right away. This improve the performance because we don't need to create `RecessOrderProperties` early, which is the type that has the bulk of the logic